pallets/liquidity-pools-gateway/queue/src/lib.rs

```diff
@@ -128,21 +121,25 @@ pub mod pallet {
 		/// are not reverted.
 		///   - an extra defensive weight is added in order to cover the weight
 		/// used when processing the message.
-		#[pallet::weight(T::WeightInfo::process_message())]
+		#[pallet::weight(MessageQueue::<T>::get(nonce)
+            .map(|msg| T::MessageProcessor::max_processing_weight(&msg))
+            .unwrap_or(T::DbWeight::get().reads(1)))]
 		#[pallet::call_index(0)]
-		pub fn process_message(origin: OriginFor<T>, nonce: T::MessageNonce) -> DispatchResult {
+		pub fn process_message(
```

Comment:
We compute the max required weight for the specific message upfront. Later the extrinsic could be consume less